mypy/subtypes.py

```diff
@@ -302,6 +331,237 @@ def visit_type_type(self, left: TypeType) -> bool:
+            assert isinstance(dec, Decorator)
+            return find_node_type(dec.var, itype, subtype)
+        return find_node_type(method, itype, subtype)
+    else:
+        # don't have such method, maybe variable or decorator?
+        node = info.get(name)
+        if not node:
+            v = None
+        else:
+            v = node.node
+        if isinstance(v, Decorator):
+            v = v.var
+        if isinstance(v, Var):
+            return find_node_type(v, itype, subtype)
+        if not v and name not in ['__getattr__', '__setattr__', '__getattribute__']:
+            for method_name in ('__getattribute__', '__getattr__'):
+                # Normally, mypy assumes that instances that define __getattr__ have all
+                # attributes with the corresponding return type. If this will produce
+                # many false negatives, then this could be prohibited for
+                # structural subtyping.
+                method = info.get_method(method_name)
+                if method and method.info.fullname() != 'builtins.object':
+                    getattr_type = find_node_type(method, itype, subtype)
+                    if isinstance(getattr_type, CallableType):
+                        return getattr_type.ret_type
+        if itype.type.fallback_to_any:
+            return AnyType(TypeOfAny.special_form)
+    return None
+
+
+def get_member_flags(name: str, info: TypeInfo) -> Set[int]:
+    """Detect whether a member 'name' is settable, whether it is an
+    instance or class variable, and whether it is class or static method.
+    """
+    method = info.get_method(name)
+    setattr_meth = info.get_method('__setattr__')
+    if method:
+        # this could be settable property
+        if method.is_property:
+            assert isinstance(method, OverloadedFuncDef)
+            dec = method.items[0]
+            assert isinstance(dec, Decorator)
+            if dec.var.is_settable_property or setattr_meth:
+                return {IS_SETTABLE}
+        return set()
+    node = info.get(name)
+    if not node:
+        if setattr_meth:
+            return {IS_SETTABLE}
+        return set()
+    v = node.node
+    if isinstance(v, Decorator):
+        if v.var.is_staticmethod or v.var.is_classmethod:
+            return {IS_CLASS_OR_STATIC}
+    # just a variable
+    if isinstance(v, Var):
+        flags = {IS_SETTABLE}
+        if v.is_classvar:
+            flags.add(IS_CLASSVAR)
+        return flags
+    return set()
+
+
+def find_node_type(node: Union[Var, FuncBase], itype: Instance, subtype: Type) -> Type:
+    """Find type of a variable or method 'node' (maybe also a decorated method).
+    Apply type arguments from 'itype', and bind 'self' to 'subtype'.
+    """
+    from mypy.checkmember import bind_self
+    if isinstance(node, FuncBase):
+        typ = function_type(node,
+                            fallback=Instance(itype.type.mro[-1], []))  # type: Optional[Type]
```

Comment:
Actually I wanted to use ``function``, but I don't know how to access it in this module without creating some import cycles. Do you have an idea for this?